### PR TITLE
Encryption in MongoDB

### LINKED ISSUE #1940: Encryption in MongoDB.
Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of article
Encryption at rest in MongoDB.

### Introduction : 
The article is about Encryption at rest in MongoDB. In recent years, MongoDB has risen to become top player in non-relational database providers in the world and we all know that it is so much necessary and important to protect our data and databases from non legitimate access. 

So my article will give thorough knowledge and information about how to encrypt MongoDB databases using encryption at rest and also explain some other related terms regarding encryption in a clear and easy way so that the readers can realise the importance of implementing security features on databases. 

### Key takeaways: 
After reading the article readers will be able to understand :

1. What encryption at rest means in MongoDB. 
2. How to implement encryption at rest. 
3. Encryption key management in MongoDB. 
4. Information on encryption life cycle. 
5. Best practices for encryption in MongoDB. 

### References: 
N/A

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by HARITJOSHI1**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes #1940

### CODE REVIEW

**Comment by HARITJOSHI1** [comment]
@katungi  Sir any update?

**Comment by katungi** [comment]
Hello @HARITJOSHI1 The preliminary review has began. Let's get this show on the road.

**Comment by HARITJOSHI1** [comment]
> Hello @HARITJOSHI1 The preliminary review has began. Let's get this show on the road.

Ohh yes!

**Comment by HARITJOSHI1** [comment]
@katungi any new update sir?

**Comment by HARITJOSHI1** [comment]
@hectorkambow what is the status? Pls reply sir

File: articles/encryption-in-mongodb/index.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+In this article, we are going to learn about Encryption at rest in MongoDB with some examples.
+
+### Prerequisites
+
+- MongoDB is installed in your system. To install MongoDB safely in your system visit the official documentation via this link
+  <https://docs.mongodb.com/manual/installation/>
+
+- Must have some prior knowledge about MongoDB and its shell commands.
+
+So before starting with the concept of Encryption at rest in MongoDB let's first discuss the term 'Encryption'.
```
Comment (katungi) [inline articles/encryption-in-mongodb/index.md:28]: Avoid using _So_

File: articles/encryption-in-mongodb/index.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+### What is a key in cryptography?
+
+A cryptographic key is a string of characters used within an encryption algorithm for altering data so that it appears random. Like a physical key, it encrypts data so that only someone with the right key can decrypt it.
+
+### Why is data encryption necessary?
+
+Data encryption is very much necessary and the factors below are the key reason to encrypt our data or databases as,
+
+1. Privacy: So with the help of encryption no one can read or interfere with the data you want to protect except the intended parties. This prevents attackers, and networks, internet service providers, and in some cases governments from intercepting and reading sensitive data.
```
Comment (katungi) [inline articles/encryption-in-mongodb/index.md:42]: This sentence would flow better without the _So_

File: articles/encryption-in-mongodb/index.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+```bash
+cat mongodb-cert.key mongodb-cert.crt > mongodb.pem
+```
+
+So after this, you will have a file named `mongodb.pem` in your disk and that is the file you will need to enable SSL encryption.
+
+Now navigate to that where you have stored your `.pem` file and there we need to set an SSL mode by using `--sslMode` argument in the command which I will show you next which defines whether the SSL is enabled or not and how to configure it.
+
+Again for more information please visit the MongoDB official documentation here <https://docs.mongodb.com/manual/tutorial/configure-ssl/>
```
Comment (katungi) [inline articles/encryption-in-mongodb/index.md:125]: It would be better without _Again_

File: articles/encryption-in-mongodb/index.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+So this command will start an SSL sever without a CA certification file for which you have to get a license from the government and that server will wait for a connection & now to connect to that server we have to pass our `mongodb.pem` file as a CA certificate to complete the process so open a new tab in your terminal window and type
+
+```bash
+mongodb --ssl --sslCAFile mongodb.pem --host localhost
+```
+
+And with this, you have created an SSL encrypted connection where you can create your database and add secure functionality to it.
+
+I hope you enjoyed reading this article as much as I enjoyed writing it.
```
Comment (katungi) [inline articles/encryption-in-mongodb/index.md:140]: You can have a more elaborate conclusion.

**Review by katungi — COMMENTED** [review]
Hey @HARITJOSHI1, Amazing article. I personally needed it.
Just a few notes:
- You used so way too many times in the article
- You can use a grammar service like grammarly or hemmingway to check how long your sentences are. Your sentences are a tad too long. Please look into that.
Overall, it is a great article

**Comment by HARITJOSHI1** [comment]
> Hey @HARITJOSHI1, Amazing article. I personally needed it.
> Just a few notes:
> - You used so way too many times in the article
> - You can use a grammar service like grammarly or hemmingway to check how long your sentences are. Your sentences are a tad too long. Please look into that.
> Overall, it is a great article 

Can you point it out in the article as it will help me to correct the mistakes faster.

**Comment by HARITJOSHI1** [comment]
@katungi  I did some changes pls review it.

**Comment by katungi** [comment]
> @katungi I did some changes pls review it.

On it @HARITJOSHI1

**Comment by HARITJOSHI1** [comment]
@katungi  Any update? Are the preliminary checks done?

**Comment by HARITJOSHI1** [comment]
@hectorkambow Any update? Are the preliminary checks done?

File: articles/encryption-in-mongodb/index.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+Before starting with the concept of Encryption at rest in MongoDB let's first discuss the term 'Encryption'.
+
+### What is Encryption?
+
+Encryption is a way of securing data so that only authorized parties or users can understand the information. In other words, it is the process of converting human-readable plaintext to incomprehensible text, also known as ciphertext.
+
+### What is a key in cryptography?
+
+A cryptographic key is a string of characters used within an encryption algorithm for altering data so that it appears random. Like a physical key, it encrypts data so that only someone with the right key can decrypt it.
```
Comment (katungi) [inline articles/encryption-in-mongodb/index.md:36]: _to alter data_ flows better.

File: articles/encryption-in-mongodb/index.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+MongoDB does not include an enterprise encryption key management solution, and users must purchase a solution from a third-party key management solution provider. MongoDB Enterprise Editions and enables customers to protect encryption using several tested and validated enterprise key management partners which is a very plus point of MongoDB.
+
+Now you may ask why key management is necessary so here is the answer,
+
+Encryption key management is necessary because without key management you left the keys to protecting your sensitive business and customer data exposed, then you expose your entire organization to the risk of data loss or theft.
+Today, with MongoDB Enterprise, customers can meet encryption and key management best practices easily through implementing native encryption and deploying a third-party enterprise key management solution.
+
+Before we moving on to how to implement SSL encryption in your database let's first understand what is SSL.
```
Comment (katungi) [inline articles/encryption-in-mongodb/index.md:95]: Change this to either _Before moving on to..._ or _Before we move one to..._

File: articles/encryption-in-mongodb/index.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+### What is SSL?
+
+SSL stands for Secure Sockets Layer, and it refers to a protocol for encrypting and securing communications that take place on the Internet.
+The purpose of SSL is to secure communications between a client and a server, but it can also secure email, VoIP, LTE, and other communications on a shaky and untrusted network.
+
+Now I will tell you how to add SSL transport encryption to your MongoDB database.
```
Comment (katungi) [inline articles/encryption-in-mongodb/index.md:102]: I prefer you use show instead of tell. You can also choose to rephrase it to, _Now we will learn how to add SSL..._

File: articles/encryption-in-mongodb/index.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+### Implementing SSL encryption in MongoDB
+
+To do that you need to enter this command in your MongoDB shell
```
Comment (katungi) [inline articles/encryption-in-mongodb/index.md:106]: Tell the reader what _that_ is. 
_To implement SSL encryption in mongoDB..._

File: articles/encryption-in-mongodb/index.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+After that process, you will end up with two files the `mongodB-cert` key file `mongodb-cert.crt` file.
+We need to concatenate these two files into one therefore you need to execute
+
+```bash
+cat mongodb-cert.key mongodb-cert.crt > mongodb.pem
+```
+This will create a file named `mongodb.pem` in your disk and that is the file you will need to enable SSL encryption.
+
+Now navigate to that where you have stored your `.pem` file and there we need to set an SSL mode by using `--sslMode` argument in the command which I will show you next which defines whether the SSL is enabled or not and how to configure it.
```
Comment (katungi) [inline articles/encryption-in-mongodb/index.md:123]: remove the word _that_, it is unnecessary

**Review by katungi — COMMENTED** [review]
Check on these final changes. 
I already did the tech review as I was going through the article this time.

File: articles/encryption-in-mongodb/index.md (hunk 8 above)
Comment (HARITJOSHI1) [reply articles/encryption-in-mongodb/index.md:106]: I already did that part

**Comment by HARITJOSHI1** [comment]
> Check on these final changes. 
> I already did the tech review as I was going through the article this time. 

I will check it out and update it and after that I will give u a reminder for the same

File: articles/encryption-in-mongodb/index.md [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+This is why MongoDB has conducted performance tests using Intel Xeon X5675 CPUs thus storage engine experiences an average latency between 10%-15%, depending on the amount of data that a user is reading or writing to the database.
+
+When the user writes only large amounts of data to the database, the performance impacts fall on the server-side, however, in more commons scenarios a user might be performing read-only commands for the data so the performance will likely fall between 5%-10%.
+
+### Encryption Key Management in MongoDB
+
+Encryption key management is the method used to protect and manage your encryption keys. Enterprise encryption key management should meet the key management framework and recommendations as suggested by NIST in Special Publications.
+MongoDB does not include an enterprise encryption key management solution, and users must purchase a solution from a third-party key management solution provider. MongoDB Enterprise Editions and enables customers to protect encryption using several tested and validated enterprise key management partners which is a very plus point of MongoDB.
+
+Now you may ask why key management is necessary so here is the answer,
```
Comment (katungi) [inline articles/encryption-in-mongodb/index.md:90]: Remove this sentence...It is not necessary.

**Review by katungi — APPROVED** [review]
With this change...Everything looks okay.
@HARITJOSHI1 Avoid using rhetorical questions in the article, they are repetitive and serve little purpose to the flow. 
You can also try reading the article from a readers perspective and be honest to yourself about what you liked and what you didn't about the article.
Finally, when explaining to the reader and directing them to where a specific task should be done, be very clear and intentional about the locations and directories or files.
It is a great article overall. It is ready for final review .

cc @hectorkambow

**Comment by HARITJOSHI1** [comment]
@katungi @hectorkambow what's the status of my article?

**Comment by hectorkambow** [comment]
> @katungi @hectorkambow what's the status of my article? @HARITJOSHI1 

NOTE: Thank you for the follow up  - we are making some changes to the main repo now - and this is delaying a few article from being published as we want to implement the changes 1st - once these are done - we will publish this one asap. Thank you 🚀

**Comment by hectorkambow** [comment]
@HARITJOSHI1  

Please go over the edits I've made during my final review. There were some area a bit difficult to get through - see the edits to apply if applicable moving forward in future articles - please be sure to check our resources page for more tips. 

One thing I recall was the in article links - you had the wrong syntax in place. 


cc @katungi  please do the same as articles should not be pushed for final review with these type of issues.

**Comment by HARITJOSHI1** [comment]
> @HARITJOSHI1  
> 
> Please go over the edits I've made during my final review. There were some area a bit difficult to get through - see the edits to apply if applicable moving forward in future articles - please be sure to check our resources page for more tips. 
> 
> One thing I recall was the in article links - you had the wrong syntax in place. 
> 
> 
> cc @katungi  please do the same as articles should not be pushed for final review with these type of issues. 
> 
> 

Which wrong syntax links are you talking about sorry but I didn't get it

**Comment by hectorkambow** [comment]
> > @HARITJOSHI1
> > Please go over the edits I've made during my final review. There were some area a bit difficult to get through - see the edits to apply if applicable moving forward in future articles - please be sure to check our resources page for more tips.
> > One thing I recall was the in article links - you had the wrong syntax in place.
> > cc @katungi  please do the same as articles should not be pushed for final review with these type of issues.
> 
> Which wrong syntax links are you talking about sorry but I didn't get it

The link syntax - when you were referring the reader to official docs or other blog posts the syntax was <> 
I believe - you can see the changes in my edits as well.

**Comment by HARITJOSHI1** [comment]
> > > @HARITJOSHI1
> > > Please go over the edits I've made during my final review. There were some area a bit difficult to get through - see the edits to apply if applicable moving forward in future articles - please be sure to check our resources page for more tips.
> > > One thing I recall was the in article links - you had the wrong syntax in place.
> > > cc @katungi  please do the same as articles should not be pushed for final review with these type of issues.
> > 
> > Which wrong syntax links are you talking about sorry but I didn't get it
> 
> The link syntax - when you were referring the reader to official docs or other blog posts the syntax was <> 
> I believe - you can see the changes in my edits as well. 
>  It should look like [alt tag] (URL)

Yeah I can see the changes but to be honest  I don't know quite sure how the syntax got this way cuz in my initial commit I didn't add any wrong syntax regarding to the links but anyways I will look into it and update the article asap.

**Comment by HARITJOSHI1** [comment]
@hectorkambow I have done the updates

**Comment by hectorkambow** [comment]
> @hectorkambow I have done the updates

I had already changed them there was no need that was a comment for future reference in future articles - are you doing git pulls to your repo (regarding our changes). 

Any changes we do should be reflected on your article - if you do a git pull.

**Comment by hectorkambow** [comment]
Published 
https://www.section.io/engineering-education/enryption-in-mongodb/
